server/json-rpc/src/lib.rs

```diff
@@ -37,86 +51,221 @@ impl fmt::Display for SubgraphRemoveParams {
+                .remove(params.name_or_id)
+                .map_err(|e| json_rpc_error(JSON_RPC_REMOVE_ERROR, e.to_string()))
+                .map(|_| Ok(Value::Null))
+                .flatten(),
+        )
+    }
+
+    /// Handler for the `subgraph_authorize` endpoint.
+    ///
+    /// Takens subgraph name and returns an auth token that can be used to
+    /// add/remove subgraphs under the input name.
+    ///
+    /// Requires bearer authorization with the master token.
+    fn authorize_handler(
+        &self,
+        params: SubgraphAuthorizeParams,
+        auth: AuthorizationHeader,
+    ) -> Result<Value, jsonrpc_core::Error> {
+        info!(self.logger, "Received subgraph_authorize request"; "params" => params.to_string());
+
+        let master_token = env::var(GRAPH_MASTER_TOKEN_VAR);
+        match &master_token {
+            Ok(master_token) if *master_token == auth.bearer_token => (), // Authorized.
+            Ok(_) => {
+                return Err(json_rpc_error(
+                    JSON_RPC_UNAUTHORIZED_ERROR,
+                    "auth token is invalid".to_owned(),
+                ))
+            }
+            Err(_) => {
+                return Err(json_rpc_error(
+                    JSON_RPC_UNAUTHORIZED_ERROR,
+                    "internal error".to_owned(),
+                ))
+            }
+        }
+        // Generate a random token of 50 ASCII alhpanumerics.
+        let mut rng = rand::thread_rng();
+        let mut token = String::new();
+        for _ in 0..50 {
+            token.push(rng.sample(Alphanumeric));
+        }
+        self.authorized
+            .write()
+            .unwrap()
+            .insert(params.name, token.clone());
+        Ok(Value::String(token))
     }
 }
 
-impl JsonRpcServerTrait for JsonRpcServer {
+impl<T: SubgraphProvider> JsonRpcServerTrait<T> for JsonRpcServer<T> {
     type Server = Server;
 
-    fn serve(
-        port: u16,
-        provider: Arc<impl SubgraphProvider>,
-        logger: Logger,
-    ) -> Result<Self::Server, io::Error> {
+    fn serve(port: u16, provider: Arc<T>, logger: Logger) -> Result<Self::Server, io::Error> {
         let addr = SocketAddrV4::new(Ipv4Addr::new(0, 0, 0, 0), port);
 
-        let mut handler = IoHandler::new();
+        let mut handler = MetaIoHandler::with_compatibility(Compatibility::Both);
 
+        let arc_self = Arc::new(JsonRpcServer {
+            provider,
+            logger,
+            authorized: Arc::new(RwLock::new(BTreeMap::new())),
+        });
         // `subgraph_deploy` handler.
-        let deploy_provider = provider.clone();
-        let deploy_logger = logger.clone();
-        let deploy_handler = move |params| {
-            Self::deploy_handler(params, deploy_provider.clone(), deploy_logger.clone())
-        };
-        handler.add_method("subgraph_deploy", move |params: Params| {
-            let deploy_handler = deploy_handler.clone();
+        let me = arc_self.clone();
```

Comment:
`me` -> `deploy_self` maybe?

Reply:
Or `self_deploy`?

Reply:
I just went with a name that could be used on all handlers so that their code looked as similar as possible. I think it looks nice and symmetrical, but if you think different names is better then your suggestion is good.

Reply:
Ok, yeah, I get that. 👍